File: front-end/src/service/review-service.ts
import http from "@/service/http";
import { ApiResponse, PageResponse } from "@/type/api-type";
import {
    ProductRatingSummary,
    ReviewRequest,
    ReviewResponse,
} from "@/type/review-type";

/**
 * Lấy danh sách đánh giá của một sản phẩm dựa trên Slug
 * @param slug Đường dẫn thân thiện của sản phẩm
 * @param params Object chứa: page, size, sortBy, sortDirection
 */
export const getReviewsByProduct = async (slug: string, params?: any) =>
    http.get<ApiResponse<PageResponse<ReviewResponse>>>(
        `/reviews/product/${slug}`,
        { params }
    );
/**
 * Lấy tóm tắt thống kê đánh giá (tổng số, trung bình sao, chi tiết từng sao)
 * @param productId ID của sản phẩm
 */
export const getReviewSummary = async (productId: number) =>
    http.get<ApiResponse<ProductRatingSummary>>(
        `/reviews/product/${productId}/summary`
    );
/**
 * Tạo một đánh giá mới
 */
export const createReview = async (data: ReviewRequest) =>
    http.post<ApiResponse<ReviewResponse>>("/reviews", data);

/**
 * Lấy tất cả đánh giá (Dành cho Admin quản lý)
 * @param params Object chứa: page, size, sortBy, sortDirection
 */
export const getAllReviewsAdmin = async (params?: any) =>
    http.get<ApiResponse<PageResponse<ReviewResponse>>>("/reviews/admin/all", {
        params,
    });

/**
 * Xóa một đánh giá
 * @param id ID của bài đánh giá
 */
export const deleteReview = async (id: number) =>
    http.delete<ApiResponse<void>>(`/reviews/${id}`);

File: back-end/src/main/java/project/back_end/service/ReviewService.java
package project.back_end.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import project.back_end.request.ReviewRequest;
import project.back_end.response.ProductRatingSummary;
import project.back_end.response.ReviewResponse;

import java.util.Map;

@Service
public interface ReviewService {

    /**
     * Tạo đánh giá mới từ người dùng
     */
    ReviewResponse createReview(ReviewRequest request, String username);

    /**
     * Lấy danh sách đánh giá của một sản phẩm (Phân trang)
     * Thường dùng ở trang chi tiết sản phẩm
     */
    Page<ReviewResponse> getReviewsByProduct(String productSlug, Pageable pageable);

    /**
     * Lấy danh sách đánh giá của một người dùng cụ thể
     * Thường dùng ở trang Profile -> Lịch sử đánh giá
     */
    Page<ReviewResponse> getReviewsByUser(Long userId, Pageable pageable);

    /**
     * Lấy tất cả đánh giá (Dành cho Admin quản lý)
     */
    Page<ReviewResponse> getAllReviews(Pageable pageable);

    /**
     * Xóa đánh giá (Admin hoặc chính chủ sở hữu)
     */
    void deleteReview(Long reviewId, String username);

    /**
     * Tính toán thống kê rating cho sản phẩm (Số lượng 5*, 4*, 3*...)
     * Trả về Map<Số sao, Số lượng>
     */
    Map<Integer, Long> getRatingStatistics(Long productId);

    /**
     * Tính điểm trung bình rating của sản phẩm
     */
    Double getAverageRating(Long productId);

    ProductRatingSummary getReviewSummary(Long productId);
}